cpp/include/mcts/SearchLog.hpp

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <core/concepts/Game.hpp>
+#include <mcts/SharedData.hpp>
+
+#include <boost/json.hpp>
+
+#include <sstream>
+#include <string>
+
+namespace mcts {
+
+template <core::concepts::Game Game>
+class Graph;
+
+template <core::concepts::Game Game>
+class SearchLog {
+ public:
+  SearchLog(const SharedData<Game>* shared_data) : shared_data_(shared_data) {}
+
+  void add_graph(const Graph<Game> graph) { graphs.push_back(graph); }
+  void build_graph(Graph<Game>& graph);
+  void update();
+  boost::json::object combine_json();
+  std::string json_str() {return boost::json::serialize(combine_json());};
+  void write_json_to_file(const boost::filesystem::path& filename);
+
+ private:
+  const SharedData<Game>* shared_data_;
+  std::vector<Graph<Game>> graphs;
+};
+
+template <core::concepts::Game Game>
+class Graph {
```

Comment:
In my opinion, `Graph` should be an inner-class of `SearchLog`. After all, `Graph` is not something that has any meaning/purpose outside of `SearchLog`. The name `Graph` is too generic as a top-level class, and that generic-ness is misleading. But as an inner-class of `SearchLog`, the name `Graph` is perfectly acceptable.

As a cosmetic benefit, you wouldn't need the forward-declaration of `Graph` at the top of the file (I personally try to avoid forward-declarations, they feel like a sign that your code is too tangled).

With this change, I would put `Node` and `Edge` as inner-classes of `SearchLog` as well (as sibling classes of `Graph`, rather than inner-classes of `Graph`).

Reply:
fixed